Replace `Merge` and `MergeNamed` with `merge()`

llucax opened the pull request:
- Remove `MergeNamed`
- Add a `merge()` wrapper function that replaces `Merge`
- Rename `Merge` to `_Merge` to make it private more explicitly
- Rename `_Merge` variable `args` to `receivers`
- Improve string representation of merge receiver
- Raise an exception is `merge()` is called with less than two receivers

Fixes #236, fixes #237.

llucax committed:
Remove `MergeNamed`

We already have 2 ways to join multiple receivers, `Merge` and `select`.
`MergeNamed` is a mixture of both, so it is redundant.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add a `merge()` wrapper function that replaces `Merge`

This is a more convenient way to use `Merge`, users normally only care
about merging some channels, so having a class just makes the syntax
more weird for most use cases that will be `async for msg in
merge(...)`.

`Merge` is not exposed to the user any more.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Rename `Merge` to `_Merge` to make it private more explicitly

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Rename `_Merge` variable `args` to `receivers`

This is just to make it more consistent with the internal attribute.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Improve string representation of merge receiver

Since `merge()` uses the `_Merge` class, when printing what's returned
by `merge()` the string representation of `_Merge` is used. This is not
very useful since it doesn't show the name that the user wrote in the
code.

This commit adds a `name` argument to `_Merge` and uses it to create the
string representation of the receiver. If no name is provided, the class
name is used as before.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Raise an exception is `merge()` is called with no receivers

Calling `merge()` with no receivers makes returns a stopped receiver,
which can be unexpected if it was done accidentally, as the user will
probably expect the merge receiver to block while receiving at least
once.

By raising an exception we make sure that users express their intentions
explicitly instead, having to write something like:

```python
receivers: list[Receiver[T]]
...
if receivers:
    async for m in merge(*receivers):
        ...
```

In this case it is clear the user is having the no-receivers case into
account and will handle it properly.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Update release notes

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

daniel-zullo-frequenz commented on src/frequenz/channels/__init__.py:
```diff
@@ -26,11 +26,10 @@
-* [Merge][frequenz.channels.Merge] and [MergeNamed][frequenz.channels.MergeNamed]:
-  [Receivers][frequenz.channels.Receiver] that merge messages coming from multiple
-  receivers into a single stream.
+* [merge][frequenz.channels.merge]: Merges messages coming from multiple receivers into
+  a single stream.
```
To be consistent with `select` documentation
```suggestion
* [merge][frequenz.channels.merge]: Merge messages coming from multiple receivers into
  a single stream.
```

daniel-zullo-frequenz commented on src/frequenz/channels/_merge.py:
```diff
@@ -13,13 +13,23 @@
+            single stream.
+
+    Raises:
+        ValueError: if no receivers are provided.
```
I'd be more precise here as it was done for `_Merge.__init__()`, otherwise it sounds to me like one receiver might be enough

```suggestion
        ValueError: if less than two receivers are provided.
```

llucax replied:
Ah, yeah, I forgot to update that docstring too, at first I checked that it wasn't empty, but then though it doesn't make much sense for only one receiver either.

daniel-zullo-frequenz commented on src/frequenz/channels/_merge.py:
```diff
@@ -29,25 +39,34 @@ class Merge(Receiver[_T]):
         Args:
-            *args: sequence of channel receivers.
+            *receivers: The receivers to merge.
+            name: The name of the receiver. Used to to create the string representation
```
```suggestion
            name: The name of the receiver. Used to create the string representation
```

daniel-zullo-frequenz reviewed (commented):
I only found a few cosmetics in the documentation to check for, otherwise LGTM

daniel-zullo-frequenz commented on src/frequenz/channels/_merge.py:
```diff
@@ -13,13 +13,23 @@
+
+    Returns:
+        A receiver that merges the messages coming from multiple receivers into a
+            single stream.
```
`The receiver` instead according to dart suggestions?

"The" also for arguments and return documentation. The rationale is that for attributes/arguments/returns you are talking about a specific instance ("this" instance), so a definite article is more appropriate, while for classes you are talking about an unspecified instance, any instance, so an indefinite article is more appropriate.

llucax replied:
Hahaha, well played Zullo, well played... (BTW, the rationale comes from me, not the dart docs, so you are sending me a citation of what I wrote myself :rofl: )

BTW, I thought about A/the here, and since it is not returning an attribute, but creating a complete new object, so every call will return a new one, not "the" same one, I thought "A" was more appropriate in this context. I only take these guides as suggestion, not law, and chose whatever I feel is more appropriate for each case.

Let me know if you still feel it should be updated in this case.

daniel-zullo-frequenz replied:
haha I thought it was taken from the dart suggestions :angel: 

That's fine with me as you have thought about it and it wasn't randomly documented. I agree those are just suggestions and there is no need to be strict with them. Though I'd prefer to keep documentation consistent whenever is possible. So far, "The" seems to be always a good fit for arguments.

llucax committed:
Apply suggestions from code review

Apply suggestions made by @daniel-zullo-frequenz.

Co-authored-by: daniel-zullo-frequenz <120166726+daniel-zullo-frequenz@users.noreply.github.com>
Signed-off-by: Leandro Lucarella <luca@llucax.com>

shsms commented on src/frequenz/channels/_merge.py:
```diff
@@ -53,7 +56,13 @@ def __init__(self, *receivers: Receiver[_T], name: str | None) -> None:
+        Raises:
+            ValueError: if less than two receivers are provided.
         """
+        if len(receivers) < 2:
+            raise ValueError("At least two receivers must be provided")
```
if it is a single component, we could actually just return the receiver, instead of a `_Merge`.  and raise only when there are 0 receivers.

that way, users can still say "get receivers for all batteries and  merge them" and have to do special case handling only for one case.

llucax replied:
True. That said, now I wonder if we shouldn't allow also empty `merges()` just for the sake of generality, so in general if people are doing something like:

```py
receivers = []
if x:
    receivers.add(x)
if y:
    receivers.add(x)
async for m in merge(*receivers):
    ...
```

Without needing to special case.

Is not exactly the same as when there is only one receiver, as if there are no receivers, the `async for` loop will end without any iterations, so I'm not 100% convinced which would be the most useful default, as users will probably need to handle the empty case specially in any case... :thinking:

shsms replied:
I agree, maybe all we need are some logs instead of exceptions.

Marenz replied:
We could have a c'tor flag or parameter `allow_empty=` or `min_required_receivers =`

llucax replied:
I'm also not convinced about the logging, as if you call `merge()` with no receivers (or one) on purpose, then you'll get your logs littered with basically noise. We could add those ctor flags to decide if we raise or emit logs but I find it overkill to be honest.

My feeling for now is:

* `merge(receiver)` should be fine, there are really no surprises there, raising or logging here is being overly patronizing, as everything will work as expected.
* `merge()` should raise. It needs special casing anyway, as it would be very weird that one wants the `merge()` loop to exit without doing any iteration, and if that's your intention, then it can be explicitly expressed using:

  ```python
  receivers = []
  if x:
      receivers.add(x)
  if y:
      receivers.add(x)
  if receivers:
      async for m in merge(*receivers):
          ...
  ```

  Which is only one extra line and makes everything very clear and explicit.

llucax replied:
OK, I went with this approach (raise with no receivers, optimize to return the same receiver if only one is provided). I also moved the raising to the `merge()` function, as `_Merge` is a low level construct. This way I can also avoid the `# noqa` for the indirectly raised exception.

shsms reviewed (commented):
Sorry about the late review, just got one comment.

llucax committed:
merge: Return the one receiver if only one receiver is provided

In case `merge()` is called with one receiver there is no need to create
a `_Merge` object, we can just return the provided receiver instead.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
merge: Move example to the top

Just to be consistent to how docstrings are normally written and
organized.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Mark merge integration test as such

The current tests for merge are really integration tests, as it uses a
channel, so we better rename it and mark it as such, as a failure in
this test could be actually a failure in `Anycast` instead.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add unit tests for the latest changes to merge

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
Updated. Sorry @shsms, I forgot to create fixup/new commits for all the changes and amended one one the old commits (to only raise when no receivers are passed). The diff for that amend is [this one](https://github.com/frequenz-floss/frequenz-channels-python/compare/d67b4d63b916efd494803ba0a1c8fe6bc5788416..f9c076f0d86f4fe2765961dabc6f4e4f9521044e).

All the other changes are in new commits.

shsms commented:
Thanks @llucax, I have no further comments, will approve once you fixup.

llucax commented:
Rebased to apply the fixups and enabled auto-merge.
